### PR TITLE
feat: Share event details along with image

### PR DESCRIPTION
**Pull request by aman210697**
Fixes: #1315 

Changes:
Share event details along with event image
Added file_provider.xml to use cache event image URI  to work with API 24+

Screenshot
WhatsApp Image 2019-03-16 at 12 58 18 AM

PS: I used placeholder image as event image for a testing purpose.

### CODE REVIEW

File: app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt [hunk 1]
```diff
@@ -195,4 +204,33 @@ object EventUtils {
...
+            val file = File(imageView.context.cacheDir, "share_image_" + System.currentTimeMillis() + ".png")
+            val fos = FileOutputStream(file)
+            bmp?.compress(Bitmap.CompressFormat.PNG, 90, fos)
+            Timber.d("file: " + file.absolutePath)
+            fos.flush()
+            fos.close()
+            bmpUri = FileProvider.getUriForFile(imageView.context, BuildConfig.APPLICATION_ID + ".provider", file)
+            Timber.d("uri: " + bmpUri)
+        } catch (e: IOException) {
+            e.printStackTrace()
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: Use Timber.e

File: app/src/main/java/org/fossasia/openevent/general/event/EventViewHolder.kt [hunk 2]
```diff
@@ -44,10 +44,14 @@ class EventViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
...
-            sendIntent.putExtra(Intent.EXTRA_TEXT, EventUtils.getSharableInfo(event))
             sendIntent.type = "text/plain"
-            itemView.context.startActivity(Intent.createChooser(sendIntent, "Share Event Details"))
+            sendIntent.putExtra(Intent.EXTRA_TEXT, EventUtils.getSharableInfo(event))
+            val bmpUri = EventUtils.getLocalBitmapUri(itemView.eventImage)
+            if (bmpUri != null) {
+                sendIntent.type = "image/*"
+                sendIntent.putExtra(Intent.EXTRA_STREAM, bmpUri)
+            }
+            itemView.eventImage.context.startActivity(Intent.createChooser(sendIntent, "Share Event Details"))
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/event/EventViewHolder.kt]: Duplicate logic. Create a util function

Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventViewHolder.kt]: @iamareebjamal sorry didn't get you.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/EventViewHolder.kt]: This logic is duplicated. Create a util function to reuse code

Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventViewHolder.kt]: done! please review @iamareebjamal

**Comment by iamareebjamal** [comment]
@aman210697 Please resolve conflicts

**Comment by aman210697** [comment]
Changes:
- Removed the share event listener.
- If image loading is successful then imageView tag will be set to String, so if imageView tag is not null then share image else only event details

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt [hunk 3]
```diff
@@ -266,7 +268,15 @@ class EventDetailsFragment : Fragment() {
             Picasso.get()
                     .load(it)
                     .placeholder(R.drawable.header)
-                    .into(rootView.eventImage)
+                    .into(rootView.eventImage, object : Callback {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: Why

File: app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt [hunk 4]
```diff
@@ -195,4 +205,51 @@ object EventUtils {
...
+    private fun getLocalBitmapUri(imageView: ImageView): Uri? {
+        val drawable = imageView.drawable
+        val bmp: Bitmap?
+        if (drawable is BitmapDrawable) {
+            bmp = drawable.bitmap
+        } else {
+            return null
+        }
+        var bmpUri: Uri? = null
+        try {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: Use try with resources

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt (hunk 3 above)
Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: To know that the image is loaded successfully, so it'll not send a placeholder image when failed.

File: app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt (hunk 4 above)
Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: done

File: app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt [hunk 5]
```diff
@@ -195,4 +204,46 @@ object EventUtils {
...
+
+    private fun getLocalBitmapUri(imageView: ImageView): Uri? {
+        val drawable = imageView.drawable
+        val bmp: Bitmap?
+        if (drawable is BitmapDrawable) {
+            bmp = drawable.bitmap
+        } else {
+            return null
+        }
+        var bmpUri: Uri? = null
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: No use of declaration here. Declare in same line as assignment

File: app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt [hunk 6]
```diff
@@ -195,4 +204,46 @@ object EventUtils {
...
+                sendIntent.type = "image/*"
+                sendIntent.putExtra(Intent.EXTRA_STREAM, bmpUri)
+            }
+        }
+        eventImage.context.startActivity(Intent.createChooser(sendIntent, "Share Event Details"))
+    }
+
+    private fun getLocalBitmapUri(imageView: ImageView): Uri? {
+        val drawable = imageView.drawable
+        val bmp: Bitmap?
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: Make it non-nullable

**Comment by aman210697** [comment]
@iamareebjamal updated

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt (hunk 3 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: Remove callback in onDestroy

File: app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt [hunk 7]
```diff
@@ -195,4 +204,46 @@ object EventUtils {
...
+                sendIntent.type = "image/*"
+                sendIntent.putExtra(Intent.EXTRA_STREAM, bmpUri)
+            }
+        }
+        eventImage.context.startActivity(Intent.createChooser(sendIntent, "Share Event Details"))
+    }
+
+    private fun getLocalBitmapUri(imageView: ImageView): Uri? {
+        val drawable = imageView.drawable
+        val bmp: Bitmap
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: Use same line for declaration and assignment

Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: it'll be inaccessible in the try block.

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt (hunk 3 above)
Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: `Picasso.get().cancelRequest(rootView.eventImage)`
will this work?

File: app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt (hunk 7 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]:
```val bmp = try {

} catch {
	null
}

if (bmp == null)
  return null
```

Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]:
```suggestion
      val drawable = imageView.drawable
        if (drawable is BitmapDrawable) {
            drawable.bitmap?.let {
                 val file = File(imageView.context.cacheDir, "share_image_" + System.currentTimeMillis() + ".png")
                 FileOutputStream(file)
                     .use { fos ->
                         it.compress(Bitmap.CompressFormat.PNG, 90, fos)
                         Timber.d("file: %s", file.absolutePath)
                         val bmpUri =
                             FileProvider.getUriForFile(imageView.context, BuildConfig.APPLICATION_ID + ".provider", file)
                         Timber.d("uri: %s", bmpUri)
                         return bmpUri
                     }
             }
        }
        return null
```
Better?

Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]:
```val drawable = imageView.drawable
        if (drawable is BitmapDrawable) {
            drawable.bitmap?.let {
                 val file = File(imageView.context.cacheDir, "share_image_" + System.currentTimeMillis() + ".png")
                 FileOutputStream(file)
                     .use { fos ->
                         it.compress(Bitmap.CompressFormat.PNG, 90, fos)
                         Timber.d("file: %s", file.absolutePath)
                         val bmpUri =
                             FileProvider.getUriForFile(imageView.context, BuildConfig.APPLICATION_ID + ".provider", file)
                         Timber.d("uri: %s", bmpUri)
                         return bmpUri
                     }
             }
        }
        return null```

Better?

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: Invert the condition

```
if (drawable is not BitmapDrawable)
    return null

..continue
```

Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: Why?
We still have to return null in the end.

File: app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt (hunk 3 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/EventDetailsFragment.kt]: Should

File: app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt (hunk 7 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: We already are returning null after the condition check. Why do we still have to return null in the end?

Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: Because of the ?.let

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]:
```
if (drawable is not BitmapDrawable)
    return null

..continue
```

Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: Screenshot from 2019-04-08 15-09-29

IDE suggestion
```
 val drawable = imageView.drawable as? BitmapDrawable ?: return null

        drawable.bitmap?.let {
            val file = File(imageView.context.cacheDir, "share_image_" + System.currentTimeMillis() + ".png")
            FileOutputStream(file)
                .use { fos ->
                    it.compress(Bitmap.CompressFormat.PNG, 90, fos)
                    Timber.d("file: %s", file.absolutePath)
                    val bmpUri =
                        FileProvider.getUriForFile(imageView.context, BuildConfig.APPLICATION_ID + ".provider", file)
                    Timber.d("uri: %s", bmpUri)
                    return bmpUri
                }
        }
        return null
```

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]:
```
if (imageView.drawable as? BitmapDrawable == null || imageView.drawable.bitmap == null)
    return null

val file = File(imageView.context.cacheDir, "share_image_" + System.currentTimeMillis() + ".png")
return FileOutputStream(file)
  .use { fos ->
    it.compress(Bitmap.CompressFormat.PNG, 90, fos)
    Timber.d("file: %s", file.absolutePath)
    val bmpUri =
    FileProvider.getUriForFile(imageView.context, BuildConfig.APPLICATION_ID + ".provider", file)
   Timber.d("uri: %s", bmpUri)
   bmpUri
 }
```

Comment (aman210697) [reply app/src/main/java/org/fossasia/openevent/general/event/EventUtils.kt]: done..

**Pull request merged**